bionty/models.py

```diff
@@ -289,97 +289,94 @@ def import_source(
         new_source = Source.filter(**unique_kwargs).one_or_none()
-        if new_source is None:
-            new_source = Source(**unique_kwargs, **add_kwargs).save()
-        if new_source.dataframe_artifact_id is not None:
+        if not new_source:
+            new_source = Source(  # type:ignore
+                **unique_kwargs,
+                currently_used=source_record.currently_used,
+                description=source_record.description,
+                url=source_record.url,
+                source_website=source_record.source_website,
+                dataframe_artifact_id=source_record.dataframe_artifact_id,
+            ).save()
+
+        # Return early if artifact already exists
+        if new_source.dataframe_artifact_id:
             return new_source
 
-        # register the dataframe artifact
-        key = f"df__{unique_kwargs.get('organism')}__{unique_kwargs.get('name')}__{unique_kwargs.get('version')}__{unique_kwargs.get('entity')}.parquet"
+        # Generate filename if not already created
+        if not parquet_filename:
+            parquet_filename, _ = encode_filenames(**unique_kwargs)
+
+        # Create artifact if needed
+        df_artifact = None
         if isinstance(df, pd.DataFrame):
+            df_artifact = ln.Artifact.from_df(df, key=parquet_filename, run=False)
+        elif source_record.url and source_record.url.startswith("s3://bionty-assets/"):
+            df_artifact = ln.Artifact(new_source.url, run=False)
+        elif isinstance(source, PublicOntology) and not source.df().empty:
             df_artifact = ln.Artifact.from_df(
-                df, key=key, _branch_code=0, run=False
-            ).save()
-        elif source.url and source.url.startswith("s3://bionty-assets/"):
-            df_artifact = ln.Artifact(new_source.url, _branch_code=0, run=False).save()
-        else:
-            try:
-                df = getattr(bt_base, source.entity)(
-                    organism=source.organism,
-                    source=source.name,
-                    version=source.version,
-                ).df()
-            except Exception as e:
-                logger.error(
-                    "please first register a DataFrame artifact and link it to source!\n"
-                    "    → artifact = ln.Artifact.from_df(df, _branch_code=0, run=False).save()\n"
-                    "    → source.dataframe_artifact = artifact\n"
-                    "    → source.save()"
-                )
-                raise ValueError from e
-            df_artifact = ln.Artifact.from_df(
-                df, key=key, _branch_code=0, run=False
-            ).save()
+                source.df(), key=parquet_filename, run=False
+            )
 
-        new_source.dataframe_artifact = df_artifact
-        new_source.save()
-        logger.important("source added!")
+        # Save artifact and update source
+        if df_artifact:
+            df_artifact.kind = "__lamindb__"
```

Comment:
This implies that `artifact.run is None`. Is this true in your case? If you'd like to populate `artifact.run`, then we should come up with a different convention. For instance, `__bionty__`.

Reply:
I just think it will then clutter the frontend logic regarding which files to hide.

Reply:
I don't want to track runs here.